core/src/main/scala/org/apache/spark/sql/delta/PreprocessTableWithDVs.scala

```diff
@@ -0,0 +1,182 @@
+ *
+ * It modifies for plan for tables with deletion vectors as follows:
+ * Before rule: <Parent Node> -> Delta Scan (key, value).
+ *   - Here we are reading `key`, `value`` columns from the Delta table
+ * After rule:
+ *   <Parent Node> ->
+ *     Project(key, value) ->
+ *       Filter (udf(__skip_row == 0) ->
+ *         Delta Scan (key, value, __skip_row)
+ *   - Here we insert a new column `__skip_row` in Delta scan. This value is populated by the
+ *     Parquet reader using the DV corresponding to the Parquet file read
+ *     (See [[DeltaParquetFileFormat]]) and it contains 0 if we want to keep the row.
+ *     The scan created also disables Parquet file splitting and filter pushdowns, because
+ *     in order to generate the __skip_row, we need to read the rows in a file consecutively
+ *     to generate the row index. This is a cost we need to pay until we upgrade to latest
+ *     Apache Spark which contains Parquet reader changes that automatically generate the
+ *     row_index irrespective of the file splitting and filter pushdowns.
+ *   - The scan created also contains a broadcast variable of Parquet File -> DV File map.
+ *     The Parquet reader created uses this map to find the DV file corresponding to the data file.
+ *   - Filter created filters out rows with __skip_row equals to 0
+ *   - And at the end we have a Project to keep the plan node output same as before the rule is
+ *     applied.
+ */
+trait PreprocessTableWithDVs extends SubqueryTransformerHelper {
+  def preprocessTablesWithDVs(plan: LogicalPlan): LogicalPlan = {
+    transformWithSubqueries(plan) {
+      case ScanWithDeletionVectors(dvScan) => dvScan
+    }
+  }
+}
+
+object ScanWithDeletionVectors {
+  def unapply(a: LogicalRelation): Option[LogicalPlan] = a match {
+    case scan @ LogicalRelation(
+            relation @ HadoopFsRelation(
+            index: TahoeFileIndex, _, _, _, format: DeltaParquetFileFormat, _), _, _, _) =>
+      dvEnabledScanFor(scan, relation, format, index)
+    case _ => None
+  }
+
+  def dvEnabledScanFor(
+      scan: LogicalRelation,
+      hadoopRelation: HadoopFsRelation,
+      fileFormat: DeltaParquetFileFormat,
+      index: TahoeFileIndex): Option[LogicalPlan] = {
+    // If the table has no DVs enabled, no change needed
+    if (!deletionVectorsReadable(index.protocol, index.metadata)) return None
+
+    require(!index.isInstanceOf[TahoeLogFileIndex],
+      "Cannot work with a non-pinned table snapshot of the TahoeFileIndex")
+
+    // If the table has no DVs enabled, no change needed
+    if (!deletionVectorsReadable(index.protocol, index.metadata)) return None
+
+    // See if the relation is already modified to include DV reads as part of
+    // a previous invocation of this rule on this table
+    if (fileFormat.hasDeletionVectorMap()) return None
+
+    // See if any files actually have a DV
+    val spark = SparkSession.getActiveSession.get
```

Comment:
nit: pass a SparkSession in if possible

Reply:
This is not easy to pass without major refactoring, given the whole convert logic is part of the extractor. Let me know if there is a way.

Reply:
Then this is fine.